File: tracker/src/test/java/testhelpers/TestPreferences.java
package testhelpers;

import android.content.SharedPreferences;
import androidx.annotation.Nullable;

import java.util.HashMap;
import java.util.Map;
import java.util.Set;


public class TestPreferences implements SharedPreferences {
    Map<String, Object> mMap = new HashMap<>();
    Editor mEditor = new TestEditor();

    @Override
    public Map<String, ?> getAll() {
        return mMap;
    }

    @Nullable
    @Override
    public String getString(String key, String defValue) {
        if (!mMap.containsKey(key)) return defValue;
        return (String) mMap.get(key);
    }

    @Nullable
    @Override
    public Set<String> getStringSet(String key, Set<String> defValues) {
        if (!mMap.containsKey(key)) return defValues;
        //noinspection unchecked
        return (Set<String>) mMap.get(key);
    }

    @Override
    public int getInt(String key, int defValue) {
        if (!mMap.containsKey(key)) return defValue;
        return (int) mMap.get(key);
    }

    @Override
    public long getLong(String key, long defValue) {
        if (!mMap.containsKey(key)) return defValue;
        return (long) mMap.get(key);
    }

    @Override
    public float getFloat(String key, float defValue) {
        if (!mMap.containsKey(key)) return defValue;
        return (float) mMap.get(key);
    }

    @Override
    public boolean getBoolean(String key, boolean defValue) {
        if (!mMap.containsKey(key)) return defValue;
        return (boolean) mMap.get(key);
    }

    @Override
    public boolean contains(String key) {
        return mMap.containsKey(key);
    }

    @Override
    public Editor edit() {
        return mEditor;
    }

    @Override
    public void registerOnSharedPreferenceChangeListener(OnSharedPreferenceChangeListener listener) {

    }

    @Override
    public void unregisterOnSharedPreferenceChangeListener(OnSharedPreferenceChangeListener listener) {

    }

    public class TestEditor implements Editor {

        @Override
        public Editor putString(String key, String value) {
            mMap.put(key, value);
            return mEditor;
        }

        @Override
        public Editor putStringSet(String key, Set<String> values) {
            mMap.put(key, values);
            return mEditor;
        }

        @Override
        public Editor putInt(String key, int value) {
            mMap.put(key, value);
            return mEditor;
        }

        @Override
        public Editor putLong(String key, long value) {
            mMap.put(key, value);
            return mEditor;
        }

        @Override
        public Editor putFloat(String key, float value) {
            mMap.put(key, value);
            return mEditor;
        }

        @Override
        public Editor putBoolean(String key, boolean value) {
            mMap.put(key, value);
            return mEditor;
        }

        @Override
        public Editor remove(String key) {
            mMap.remove(key);
            return mEditor;
        }

        @Override
        public Editor clear() {
            mMap.clear();
            return mEditor;
        }

        @Override
        public boolean commit() {
            return true;
        }

        @Override
        public void apply() {

        }
    }
}

File: tracker/src/test/java/testhelpers/TestPreferences.kt
package testhelpers

import android.content.SharedPreferences
import android.content.SharedPreferences.OnSharedPreferenceChangeListener

class TestPreferences : SharedPreferences {
    var map: MutableMap<String, Any?> = HashMap()
    var editor: SharedPreferences.Editor = TestEditor()

    override fun getAll(): Map<String, *> {
        return map
    }

    override fun getString(key: String, defValue: String?): String? {
        if (!map.containsKey(key)) return defValue
        return map[key] as String?
    }

    override fun getStringSet(key: String, defValues: Set<String>?): Set<String>? {
        if (!map.containsKey(key)) return defValues
        return map[key] as Set<String>?
    }

    override fun getInt(key: String, defValue: Int): Int {
        if (!map.containsKey(key)) return defValue
        return map[key] as Int
    }

    override fun getLong(key: String, defValue: Long): Long {
        if (!map.containsKey(key)) return defValue
        return map[key] as Long
    }

    override fun getFloat(key: String, defValue: Float): Float {
        if (!map.containsKey(key)) return defValue
        return map[key] as Float
    }

    override fun getBoolean(key: String, defValue: Boolean): Boolean {
        if (!map.containsKey(key)) return defValue
        return map[key] as Boolean
    }

    override fun contains(key: String): Boolean {
        return map.containsKey(key)
    }

    override fun edit(): SharedPreferences.Editor {
        return editor
    }

    override fun registerOnSharedPreferenceChangeListener(listener: OnSharedPreferenceChangeListener) {
    }

    override fun unregisterOnSharedPreferenceChangeListener(listener: OnSharedPreferenceChangeListener) {
    }

    inner class TestEditor : SharedPreferences.Editor {
        override fun putString(key: String, value: String?): SharedPreferences.Editor {
            map[key] = value
            return editor
        }

        override fun putStringSet(key: String, values: Set<String>?): SharedPreferences.Editor {
            map[key] = values
            return editor
        }

        override fun putInt(key: String, value: Int): SharedPreferences.Editor {
            map[key] = value
            return editor
        }

        override fun putLong(key: String, value: Long): SharedPreferences.Editor {
            map[key] = value
            return editor
        }

        override fun putFloat(key: String, value: Float): SharedPreferences.Editor {
            map[key] = value
            return editor
        }

        override fun putBoolean(key: String, value: Boolean): SharedPreferences.Editor {
            map[key] = value
            return editor
        }

        override fun remove(key: String): SharedPreferences.Editor {
            map.remove(key)
            return editor
        }

        override fun clear(): SharedPreferences.Editor {
            map.clear()
            return editor
        }

        override fun commit(): Boolean {
            return true
        }

        override fun apply() {
        }
    }
}
